firedrake/slate/slac/compiler.py

```diff
@@ -642,36 +668,59 @@ def parenthesize(arg, prec=None, parent=None):
+    :arg out_name: name of the the final (output) GEM variable
+                   matching the last variable in the loopy kernel
+    :arg matfree: needed for seperating some expressions in gem2slate,
+                  can probably be dropped after FIXME is fixed
+
+    :return (slate_loopy, ctx): 2-tuple of loopy kernel for slate operations,
+                                and a ctx which most importantly contains gem->loopy mappings
+    :return output_variable:    a loopy GlobalArg for the output variable
     """
     # Creation of return variables for outer loopy
     shape = gem_expr.shape if len(gem_expr.shape) != 0 else (1,)
     idx = make_indices(len(shape))
     indexed_gem_expr = gem.Indexed(gem_expr, idx)
-
-    output_loopy_arg = loopy.GlobalArg("output", shape=shape,
+    output_loopy_arg = loopy.GlobalArg(out_name, shape=shape,
                                        dtype=scalar_type,
                                        is_input=True,
-                                       is_output=True)
-    args = [output_loopy_arg] + [loopy.GlobalArg(var.name, shape=var.shape, dtype=scalar_type)
-                                 for var in var2terminal.keys()]
-    ret_vars = [gem.Indexed(gem.Variable("output", shape), idx)]
+                                       is_output=True,
+                                       dim_tags=None, strides=loopy.auto, order="C")
+    args = [output_loopy_arg]
+    for var, terminal in gem2slate.items():
+        # From the gem2slate dict we only want to append vector-shaped args
+        # which are coming from AssembledVectors to the global args of the Slate kernel,
+        # meaning we don't want to append anything matrix shaped and also no solves or actions
+        if hasattr(var, "name"):
+            if not terminal.terminal or (terminal.rank > 1 and matfree):
+                t_shape = var.shape if var.shape else (1,)
+                args.append(loopy.TemporaryVariable(var.name, shape=t_shape, dtype=scalar_type, address_space=loopy.AddressSpace.LOCAL))
+            else:
+                args.append(loopy.GlobalArg(var.name, shape=var.shape, dtype=scalar_type,
+                                            is_input=True, is_output=False, dim_tags=None, strides=loopy.auto, order="C"))
 
     preprocessed_gem_expr = impero_utils.preprocess_gem([indexed_gem_expr])
 
     # glue assignments to return variable
-    assignments = list(zip(ret_vars, preprocessed_gem_expr))
+    assignments = list(zip([gem.Indexed(gem.Variable(out_name, shape), idx)], preprocessed_gem_expr))
 
     # Part A: slate to impero_c
     impero_c = impero_utils.compile_gem(assignments, (), remove_zeros=False)
 
     # Part B: impero_c to loopy
     output_arg = OutputKernelArg(output_loopy_arg)
-    return generate_loopy(impero_c, args, scalar_type, "slate_loopy", [], log=PETSc.Log.isActive()), output_arg
+    knl_name = knl_prefix + "_knl_%d" % knl_counter()  # auto generate a unique name
```

Comment:
Try to use loopy naming infrastructure, note that naming is for the function inside the file and only needs to be unique to the file